velox/functions/lib/tests/TDigestTest.cpp

```diff
@@ -0,0 +1,384 @@
+  ASSERT_THROW(digest.add(positions, 1, 0), VeloxRuntimeError);
+  ASSERT_THROW(digest.estimateQuantile(1.1), VeloxRuntimeError);
+}
+
+TEST(TDigestTest, unalignedSerialization) {
+  constexpr int N = 1e4;
+  TDigest digest;
+  std::vector<int16_t> positions;
+  for (int i = 0; i < N; ++i) {
+    digest.add(positions, i);
+  }
+  digest.compress(positions);
+  ASSERT_NEAR(digest.sum(), 1.0 * N * (N - 1) / 2, kSumError);
+  std::string buf(1 + digest.serializedByteSize(), '\0');
+  for (int offset = 0; offset < 2; ++offset) {
+    SCOPED_TRACE(fmt::format("offset={}", offset));
+    digest.serialize(buf.data() + offset);
+    TDigest digest2;
+    digest2.mergeDeserialized(positions, buf.data() + offset);
+    digest2.compress(positions);
+    for (auto q : kQuantiles) {
+      auto v = digest2.estimateQuantile(q);
+      ASSERT_NEAR(v / (N - 1), q, kRankError);
+    }
+  }
+}
+
+TEST(TDigestTest, mergeEmpty) {
+  std::vector<int16_t> positions;
+  TDigest<> digests[2];
+  std::string buf(digests[1].serializedByteSize(), '\0');
+  digests[1].serialize(buf.data());
+  digests[0].mergeDeserialized(positions, buf.data());
+  digests[0].compress(positions);
+  ASSERT_EQ(digests[0].sum(), 0);
+  ASSERT_TRUE(std::isnan(digests[0].estimateQuantile(0.5)));
+  digests[0].add(positions, 1.0);
+  digests[0].compress(positions);
+  ASSERT_EQ(digests[0].sum(), 1);
+  ASSERT_EQ(digests[0].estimateQuantile(0.5), 1);
+  digests[0].mergeDeserialized(positions, buf.data());
+  digests[0].compress(positions);
+  ASSERT_EQ(digests[0].sum(), 1);
+  ASSERT_EQ(digests[0].estimateQuantile(0.5), 1);
+}
+
+TEST(TDigestTest, deserializeJava) {
+  std::vector<int16_t> positions;
+  {
+    SCOPED_TRACE(
+        "select to_base64(cast(tdigest_agg(x) as varbinary)) from (values (2.0)) as t(x)");
+    auto data = decodeBase64(
+        "AQAAAAAAAAAAQAAAAAAAAABAAAAAAAAAAEAAAAAAAABZQAAAAAAAAPA/AQAAAAAAAAAAAPA/AAAAAAAAAEA=");
+    TDigest digest;
+    digest.mergeDeserialized(positions, data.data());
+    digest.compress(positions);
+    ASSERT_EQ(digest.compression(), tdigest::kDefaultCompression);
+    ASSERT_EQ(digest.sum(), 2.0);
+    for (auto q : kQuantiles) {
+      ASSERT_EQ(digest.estimateQuantile(q), 2.0);
+    }
+  }
+  {
+    SCOPED_TRACE(
+        "select to_base64(cast(tdigest_agg(x, w, c) as varbinary)) from (values (2.0, 2, 200.0)) as t(x, w, c)");
+    auto data = decodeBase64(
+        "AQAAAAAAAAAAQAAAAAAAAABAAAAAAAAAEEAAAAAAAABpQAAAAAAAAABAAQAAAAAAAAAAAABAAAAAAAAAAEA=");
+    TDigest digest;
+    digest.setCompression(200);
+    digest.mergeDeserialized(positions, data.data());
+    digest.compress(positions);
+    ASSERT_EQ(digest.compression(), 200);
+    ASSERT_EQ(digest.sum(), 4.0);
+    for (auto q : kQuantiles) {
+      ASSERT_EQ(digest.estimateQuantile(q), 2.0);
+    }
+  }
+  {
+    SCOPED_TRACE(
+        "select to_base64(cast(tdigest_agg(x) as varbinary)) from unnest(sequence(0, 1000)) as t(x)");
```

Comment:
nit: Maybe also test `... from unnest(shuffle(sequence(0, 1000)))...` ?

Reply:
We got the exactly same binary from this

Reply:
@Yuhta It might be because the default compression is too large? I saw this query returns different binary with and without shuffle().
```
select to_base64(cast(tdigest_agg(cast(x as double), 1, 0.01) as varbinary)) from unnest(shuffle(sequence(0, 1000))) as t(x)
```

Reply:
Yeah compression 0.01 (actually 10 in the implementation) would not satisfy the test error bound though. Also it's preferred to have something deterministic in this case. Let me add one still for sequence in order, but with a smaller compression that still satisfy the error bound.